File: app/src/main/java/com/kieran/winnipegbus/Adapters/StopTimeAdapter.java
package com.kieran.winnipegbus.Adapters;

import android.app.Activity;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import com.kieran.winnipegbus.Activities.BaseActivity;
import com.kieran.winnipegbus.R;
import com.kieran.winnipegbusbackend.ScheduledStop;
import com.kieran.winnipegbusbackend.TransitApiManager;

import java.util.List;

public class StopTimeAdapter extends ArrayAdapter<ScheduledStop> {
    private Context context;
    private int layoutResourceId;
    private boolean use24hrTime;
    private LayoutInflater inflater;
    private List<ScheduledStop> scheduledStops;

    public StopTimeAdapter(Context context, int layoutResourceId, List<ScheduledStop> scheduledStops) {
        super(context, layoutResourceId, scheduledStops);
        this.layoutResourceId = layoutResourceId;
        this.context = context;
        this.scheduledStops = scheduledStops;
        inflater = ((Activity)context).getLayoutInflater();

        loadTimeSetting();
    }

    @Override
    public View getView(int position, View row, ViewGroup parent) {
        StopTimeHolder holder;

        if(row == null) {
            row = inflater.inflate(layoutResourceId, parent, false);

            holder = new StopTimeHolder();
            holder.routeNumber = (TextView)row.findViewById(R.id.route_number_text);
            holder.routeVariantName = (TextView)row.findViewById(R.id.route_name_text);
            holder.timeStatus = (TextView)row.findViewById(R.id.time_status_text);
            holder.departureTime = (TextView)row.findViewById(R.id.departure_time_text);

            row.setTag(holder);
        } else {
            holder = (StopTimeHolder)row.getTag();
        }

        ScheduledStop scheduledStop = scheduledStops.get(position);
        holder.routeNumber.setText(Integer.toString(scheduledStop.getRouteNumber()));
        BaseActivity.setTextViewColour(context, holder.routeNumber, scheduledStop);
        holder.routeVariantName.setText(scheduledStop.getRouteVariantName());
        holder.timeStatus.setText(scheduledStop.getTimeStatus());
        holder.departureTime.setText(scheduledStop.getEstimatedDepartureTime().toFormattedString(TransitApiManager.lastQueryTime, use24hrTime));

        return row;
    }

    public void loadTimeSetting() {
        use24hrTime = BaseActivity.getTimeSetting(context);
    }

    private static class StopTimeHolder {
        TextView routeNumber;
        TextView routeVariantName;
        TextView timeStatus;
        TextView departureTime;
    }
}

File: app/src/main/java/com/kieran/winnipegbus/adapters/StopTimeAdapter.kt
package com.kieran.winnipegbus.adapters

import android.app.Activity
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView

import com.kieran.winnipegbus.activities.BaseActivity
import com.kieran.winnipegbus.R
import com.kieran.winnipegbus.views.RouteNumberTextView
import com.kieran.winnipegbusbackend.ScheduledStop
import com.kieran.winnipegbusbackend.TransitApiManager
import android.util.TypedValue
import android.widget.ImageView


class StopTimeAdapter(context: Context, private val layoutResourceId: Int, private val scheduledStops: List<ScheduledStop>) : ArrayAdapter<ScheduledStop>(context, layoutResourceId, scheduledStops) {
    private var use24hrTime: Boolean = false
    private val inflater: LayoutInflater = (context as Activity).layoutInflater

    init {
        loadTimeSetting()
    }

    override fun getView(position: Int, row: View?, parent: ViewGroup): View {
        var row = row
        val holder: StopTimeHolder

        if (row == null) {
            row = inflater.inflate(layoutResourceId, parent, false)

            holder = StopTimeHolder()
            holder.routeNumber = row!!.findViewById<View>(R.id.route_number_text) as RouteNumberTextView
            holder.routeVariantName = row.findViewById<View>(R.id.route_name_text) as TextView
            holder.hasBikeRackIcon = row.findViewById<View>(R.id.bus_bike_rack_indicator) as ImageView
            holder.hasWifiIcon = row.findViewById<View>(R.id.bus_wifi_indicator) as ImageView
            holder.isTwoBusIcon = row.findViewById<View>(R.id.two_bus_indicator) as ImageView
            holder.timeStatus = row.findViewById<View>(R.id.time_status_text) as TextView
            holder.departureTime = row.findViewById<View>(R.id.departure_time_text) as TextView

            row.tag = holder
        } else {
            holder = row.tag as StopTimeHolder
        }

        val scheduledStop = scheduledStops[position]
        holder.routeNumber?.text = String.format("%d", scheduledStop.routeNumber)
        holder.routeNumber?.setColour(scheduledStop)
        holder.routeVariantName?.text = scheduledStop.routeVariantName
        holder.hasBikeRackIcon?.visibility = if (scheduledStop.hasBikeRack) View.VISIBLE else View.GONE
        holder.hasWifiIcon?.visibility = if (scheduledStop.hasWifi) View.VISIBLE else View.GONE
        holder.isTwoBusIcon?.visibility = if (scheduledStop.isTwoBus) View.VISIBLE else View.GONE
        holder.timeStatus?.text = scheduledStop.timeStatus

        val timeText: String
        if(scheduledStop.isCancelled) {
            timeText = scheduledStop.scheduledDepartureTime!!.toFormattedString(null, use24hrTime)
            val params = holder.timeStatus!!.layoutParams as ViewGroup.LayoutParams
            params.width = spToPx(96f, context)
            holder.timeStatus!!.layoutParams = params
        }else {
            timeText = scheduledStop.estimatedDepartureTime!!.toFormattedString(TransitApiManager.lastQueryTime, use24hrTime)
        }

        holder.departureTime!!.text = timeText

        return row
    }

    fun loadTimeSetting() {
        use24hrTime = BaseActivity.getTimeSetting(context)
    }

    fun spToPx(sp: Float, context: Context): Int {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, sp, context.resources.displayMetrics).toInt()
    }

    private class StopTimeHolder {
        internal var routeNumber: RouteNumberTextView? = null
        internal var routeVariantName: TextView? = null
        internal var hasBikeRackIcon: ImageView? = null
        internal var hasWifiIcon: ImageView? = null
        internal var isTwoBusIcon: ImageView? = null
        internal var timeStatus: TextView? = null
        internal var departureTime: TextView? = null
    }
}
